src/span.h

```diff
@@ -42,18 +67,59 @@ class Span
     /** Default assignment operator. */
     Span& operator=(const Span& other) noexcept = default;
 
+    /** Construct a Span from an array. This matches the corresponding C++20 std::span constructor. */
+    template <int N>
+    constexpr Span(C (&a)[N]) noexcept : m_data(a), m_size(N) {}
```

Comment:
Could you help me understand the lifetime of a temporary here? It seems like you'd want this to be lvalue only.
For example, this works and passes undefined/memory sanitizer checks, but I don't understand why:
```c++
Span<const char> foo("bar");
printf("foo: %s\n", foo.data());
```
I would have expected the lifetime of the literal to end after the constructor, but the memory appears to remain valid afterwards.

Reply:
That's a good question, but not the right example. String literals have static lifetime in C++, so in this specific case there is no problem.

However, there are pitfalls around constructing a Span from a temporary, see https://github.com/bitcoin/bitcoin/pull/18468#issuecomment-623030711 for more discussion. It's an inherent problem with objects that store references unfortunately, and present in `std::span` as well.

Reply:
Follow-up: https://github.com/bitcoin/bitcoin/pull/19367